security/password_migration.rst

```diff
@@ -0,0 +1,212 @@
+using the ``migrate_from`` option:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/security.yaml
+        security:
+            # ...
+
+            encoders:
+                legacy:
+                    algorithm: sha256
+                    encode_as_base64: false
+                    iterations: 1
+
+                App\Entity\User:
+                    # the new encoder, along with its options
+                    algorithm: sodium
+                    migrate_from:
+                        - bcrypt # uses the "bcrypt" encoder with the default options
+                        - legacy # uses the "legacy" encoder configured above
+
+    .. code-block:: xml
+
+        <!-- config/packages/security.xml -->
+        <?xml version="1.0" encoding="UTF-8"?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:security="http://symfony.com/schema/dic/security"
+            xsi:schemaLocation="http://symfony.com/schema/dic/security
+                https://symfony.com/schema/dic/security/security-1.0.xsd">
+
+            <security:config>
+                <!-- ... -->
+
+                <security:encoder class="legacy"
+                    algorithm="sha256"
+                    encode-as-base64="false"
+                    iterations="1"
+                />
+
+                <!-- algorithm: the new encoder, along with its options -->
+                <security:encoder class="App\Entity\User"
+                    algorithm="sodium"
+                >
+                    <!-- uses the bcrypt encoder with the default options -->
+                    <security:migrate-from>bcrypt</security:migrate-from>
+
+                    <!-- uses the legacy encoder configured above -->
+                    <security:migrate-from>legacy</security:migrate-from>
+                </security:encoder>
+            </security:config>
+        </container>
+
+    .. code-block:: php
+
+        // config/packages/security.php
+        $container->loadFromExtension('security', [
+            // ...
+
+            'encoders' => [
+                'legacy' => [
+                    'algorithm' => 'sha256',
+                    'encode_as_base64' => false,
+                    'iterations' => 1,
+                ],
+
+                'App\Entity\User' => [
+                    // the new encoder, along with its options
+                    'algorithm' => 'sodium',
+                    'migrate_from' => [
+                        'bcrypt', // uses the "bcrypt" encoder with the default options
+                        'legacy', // uses the "legacy" encoder configured above
+                    ],
+                ],
+            ],
+        ]);
+
+.. tip::
+
+    The *auto* and *native* encoders automatically enables password migration
+    using the following list of algorithms:
```

Comment:
@wouterj A MigratingPasswordEncoder is automatically configured when the `migrate_from` option is not set __and__ the `algorithm` option is set to:
- auto:  The best encoder is auto-determined to SodiumPasswordEncoder _if supported_, which produces `argon2id` or `argon2i` hashes (depends on the version of the installed libsodium php extension),
If SodiumPasswodEncoder is not supported, it will be `NativePasswordEncoder`, which produces either `argon2id`, `argon2i` or `bcrypt` passwords (depends on the php version).
- argon2id: SodiumPasswordEncoder is used as best encoder if it is supported and we are sure that it can only produce argon2id hashes, otherwise NativePasswordEncoder is used with `argon2id` as fixed algorithm.
- argon2i: The best encoder resolution is the same as for argon2id above
- bcrypt: Best encoder will be `NativePasswordEncoder` with `bcrypt` as fixed algorithm.
- native: Best encoder will be `NativePasswordEncoder` and produce either `argon2id`, `argon2i` or `bcrypt` passwords depending on the php version
- sodium: Best encoder will be SodiumPasswordEncoder (must be supported), which produces `argon2id` or `argon2i` hashes depending on the libsodium php ext version.

And for all those, the extra encoders will be: `Pbkdf2PasswordEncoder` using the configured `hash_algorithm` (defaults to `sha512`) and `MessageDigestPasswordEncoder` using the configured `hash_algorithm` as well. 

Additional notes:
The `PasswordEncoderInterface` implementation used as best encoder may change as time goes by __only if__  you use `auto`.

The hashing algorithm used may change if you use one of `native` and `sodium` as value for `algorithm`, because it will respectively lead to `NativePasswordEncoder` and `SodiumPasswordEncoder` without fixed algo (so they guess the best one).

Important: both encoders will always be able to validate hashes computed using all the algorithms they support, which means that for instance using `algorithm: native` may start producing `argon2id` hashes and keep validating bcrypt hashes, allowing for opportunistic upgrades as documented below. Same applies to `algorithm: sodium`.

IMO Using `migrate_from` should be preferred over setting `hash_algorithm` when `algorithm: sodium|native|argon2id|argon2i|bcrypt` (so `hash_algorithm should only be used with `auto` when needed), it feels better and more structured to me. 

I hope I have been clear enough! :) 
I think the tip is pretty good as-is, it just needs to mention that `argon2id|argon2i|bcrypt` are also auto-migrating (any native/sodium supported algo actually)